common/workunit/workflow.cpp

```diff
@@ -613,12 +808,888 @@ WorkflowMachine::WorkflowMachine(const IContextLogger &_logctx)
+                    item.setState(WFStateDone);
+                break;
+            case WFTypeSuccess:
+                success = true;
+                //fall through
+            case WFTypeFailure:
+                item.setState(WFStateNull);
+                //This item must be the last item in the contingency to execute
+                endContingency();
+                if(checkIfDone())
+                    return;
+                //A contingency cannot have its own contingency
+                processDependentSuccessors(queryWorkflowItem(item.queryContingencyFor()));
+                if(success)
+                    processLogicalSuccessors(queryWorkflowItem(item.queryContingencyFor()));
+                return;
+            case WFTypeRecovery:
+                item.setState(WFStateSkip);
+                break;
+            }
+            unsigned successWfid = item.querySuccess();
+            if(successWfid)
+            {
+                startContingency();
+                CCloneWorkflowItem & successActivator = queryWorkflowItem(successWfid);
+                successActivator.setAbortFlag(item.queryAbortFlag());
+                successActivator.setContingencyWithin(item.queryWfid());
+                processLogicalSuccessors(successActivator);
+                return;
+            }
+        }
+        catch(WorkflowException const * e)
+        {
+            Owned<const WorkflowException> savedException = e;
+            bool hasContingency = handleFailureParallel(item, e, false);
+            //If the contingency exists, it must be fully performed before processSuccessors is called on the current item
+            //Until the clause finishes, any items dependent on the current item shouldn't execute.
+            if(hasContingency)
+                return;
+        }
+    }
+    bool inDate;
+    // NOTE - it is unlikely that this comparison will become out of date, but it is accounted for, above, with: if(abort != item.queryAbortFlag())
+    {
+        CriticalBlock thisBlock(abortCritSec);
+        inDate = (abort == item.queryAbortFlag());
+    }
+    if(inDate && !done)
+    {
+        processLogicalSuccessors(item);
+        processDependentSuccessors(item);
+    }
+#ifdef TRACE_WORKFLOW
+    LOG(MCworkflow, "Done workflow item %u", wfid);
+#endif
+}
+
+void WorkflowMachine::doExecuteItemParallel(IRuntimeWorkflowItem & item)
+{
+    try
+    {
+        performItemParallel(item.queryWfid());
+    }
+    catch(WorkflowException * ein)
+    {
+        if (ein->queryWfid() == 0)
+        {
+            StringBuffer msg;
+            ein->errorMessage(msg);
+            WorkflowException * newException = new WorkflowException(ein->errorCode(), msg.str(), item.queryWfid(), ein->queryType(), ein->errorAudience());
+            ein->Release();
+            ein = newException;
+        }
+
+        if(ein->queryType() == WorkflowException::ABORT)
+            throw ein;
+
+        //if(!attemptRetry(item, 0, scheduledWfid))
```

Comment:
Add a comment
//recovery will be added in a subsequent PR (and a reference to the jira HPCC-xxxxx)